### PR TITLE
merge upstream changes into cwise_ops_gradients.h

### PR DESCRIPTION
**Pull request by jeffdaily**

### CODE REVIEW

**Comment by whchung** [comment]
it seems quite a few breakages due to this PR...

**Comment by jeffdaily** [comment]
That's what I was hoping to ascertain.  Our version of this header has diverged slightly from upstream.  As part of https://github.com/tensorflow/tensorflow/pull/29372, I did not attempt to merge this header.  Perhaps we are blocked by the complex types not yet being available in all ops yet.  I will need to examine the build logs to know for sure.

**Comment by jeffdaily** [comment]
The errors are the following two, but repeated many times:

```
/opt/rocm/include/hip/hcc_detail/hip_vector_types.h:466:41: error: no viable overloaded '^='
        return HIP_vector_type<T, n>{x} ^= y;
               ~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/Core/GenericPacketMath.h:211:51: note: in instantiation of function template specialization 'operator^<float, 4, nullptr>' requested here
pxor(const Packet& a, const Packet& b) { return a ^ b; }
```
```
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/Core/GenericPacketMath.h:215:57: error: invalid argument type 'const HIP_vector_type<float, 4>' to unary expression
pandnot(const Packet& a, const Packet& b) { return a & (~b); }
                                                        ^~
```

**Comment by whchung** [comment]
That’s related to how vector types are implemented in HIP then. It doesnt seem easy to replace the implementation though.

File: tensorflow/core/kernels/cwise_ops_gradients.h [hunk 1]
```diff
@@ -75,14 +75,19 @@ struct scalar_inverse_gradient_op {
...
+      const T out_conj = numext::conj(output);
+      return -out_conj * out_conj * output_gradient;
+    }
   }
   template <typename Packet>
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE const Packet
   packetOp(const Packet& output, const Packet& output_gradient) const {
     const Packet out_conj = pconj(output);
-    return pnegate(pmul(output_gradient, pmul(out_conj, out_conj)));
+    return mul_no_nan_op<T>().packetOp(pnegate(pmul(out_conj, out_conj)),
```
Comment (deven-amd) [inline tensorflow/core/kernels/cwise_ops_gradients.h]: I think you will need to set PacketAccess to False for TENSORFLOW_USE_ROCM at line 97 to get around the error you are seeing. Doing so should disable this routine from getting invoked.

File: tensorflow/core/kernels/cwise_ops_gradients.h [hunk 2]
```diff
@@ -99,15 +104,20 @@ struct scalar_sqrt_gradient_op {
...
+      return (static_cast<T>(0.5) * output_gradient) / out_conj;
+    }
   }
   template <typename Packet>
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE const Packet
   packetOp(const Packet& output, const Packet& output_gradient) const {
     const Packet const_half = pset1<Packet>(static_cast<T>(0.5));
     const Packet out_conj = pconj(output);
-    return pdiv(pmul(const_half, output_gradient), out_conj);
+    return mul_no_nan_op<T>().packetOp(pdiv(const_half, out_conj),
```
Comment (deven-amd) [inline tensorflow/core/kernels/cwise_ops_gradients.h]: same comment as above

File: tensorflow/core/kernels/cwise_ops_gradients.h [hunk 3]
```diff
@@ -124,17 +134,24 @@ struct scalar_rsqrt_gradient_op {
...
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE const Packet
   packetOp(const Packet& output, const Packet& output_gradient) const {
     const Packet const_half = pset1<Packet>(static_cast<T>(-0.5));
     const Packet out_conj = pconj(output);
-    return pmul(const_half, pmul(pmul(output_gradient, out_conj),
-                                 pmul(out_conj, out_conj)));
+    auto safe_pmul = [](const Packet& a, const Packet& b) {
+      return mul_no_nan_op<T>().packetOp(a, b);
+    };
+    return safe_pmul(pmul(const_half, pmul(out_conj, out_conj)),
```
Comment (deven-amd) [inline tensorflow/core/kernels/cwise_ops_gradients.h]: same comment as above

**Pull request merged**